### ISSUE TITLE
VSCode TreeDataProvider does not reload icons with same URI after they change on disk and a refresh call is made.

### ISSUE THREAD

**Issue by tomsaunders-code** [microsoft/vscode#85501]
I'm developing a View extensions with the TreeDataProvider. Users are able to right click on an item in the tree, opening the showOpenDialog window. There they can pick an svg or png image file to replace the current icon used for the selected Tree Item. Once chosen the new image overwrites the current icon file with the same name in the same path (same URI). Then the TreeDataProvider is refreshed using the onDidChangeTreeData event. On refresh the new icon does not load. Icons with different URI's will reload. The new icon only appears after a window reload. Am I missing a call in the API to force reload the URI, otherwise this would seem like a bug.

Version: 1.41.0-insider (user setup)
Commit: a747d264d668a947eeacd3492b60df6ce8dee6b4
Date: 2019-11-23T11:09:42.834Z
Electron: 6.1.5
Chrome: 76.0.3809.146
Node.js: 12.4.0
V8: 7.6.303.31-electron.0
OS: Windows_NT x64 10.0.19025

Steps to Reproduce:

1. Using the TreeDataProvider API load a custom View into Explorer with icons
2. Change out one of the icons on disk with another icon of the same name and path.
3. Refresh the TreeDataProvider without a window reload.
4. Note the icon with same URI will not change.

Does this issue occur when all extensions are disabled?:
This issue relates to an extension feature, therefore could not be tested without extensions.

**Comment by alexr00** [comment]
@tomsaunders-code do you have some sample code a repo I could clone to make testing/verifying easier?

**Comment by tomsaunders-code** [comment]
@alexr00 absolutely, checkout master on the extension repo:

https://gitlab.com/tomsaunders-tools/vscode-workspace-explorer

The file doing the heavy lifting:
*  `/src/workspaceTreeItem.js` - Tree Item and functions to find images on treeDataProvider refresh event.

To reproduce you can do the following in this extension:
* Set the workspaceExplorer.workspaceStorageDirectory configuration
* Add/save a workspace to the directory specified in the configuration
* You should now have a `.code-workspace` file in your  workspaceExplorer.workspaceStorageDirectory directory
* Refresh the extension's explorer view
* Right click on the workspace you created
* Choose change icon and pick a png or svg. Do not reload the window.
* The extension's explorer view should refresh the tree and you'll see the icon next to your workspace name
* Right click and change the icon again or open up the workspaceExplorer.workspaceStorageDirectory and save a different image file as the same name as your workspace (overwriting the first image you saved with the right click menu)
* Hit the extensions explorer view refresh button
* Note the icon for your workspace does not change as expected

Thanks!

**Comment by alexr00** [comment]
I did a bit of learning, and this is actually intentional(and normal for browsers). If we reloaded icons from disk ever time we rendered something, we could take a performance hit depending on how many icons there are. However, there is a work around for you.

Your extension returns strings for the `light` and `dark` properties of the `iconPath`. If instead you return a `Uri` and have a `query` that changes when you know that your icon changed, then it will be loaded from disk again. As a very silly example:
```typescript
let x = 0;

const getCustomWorkspaceIcons = function getCustomWorkspaceIcons(
  workspaceFileNameAndFilePath,
  collapsableState,
  extensionConfig,
) {
  ...
    return {
      light: vscode.Uri.file(svg).with({query: `x=${x++}`}),
      dark: vscode.Uri.file(svg).with({query: `x=${x++}`}),
    };
  } 
```
Clearly you wouldn't want to change the query every time (otherwise your extension could perform worse), only when you know the icon changed.
